lib/bootloader_setup.pm

```diff
@@ -56,6 +58,91 @@ sub stop_grub_timeout {
     send_key 'up';
 }
 
+=head2 add_custom_grub_entries
+
+Add custom grub entries with extra kernel parameters.
+It adds 3rd line with default options + 4th line with advanced options.
+Extra kernel parameters are taken in GRUB_PARAM variable.
+
+e.g.  grub entries before:
+* SLES 15
+* Advanced options for SLES 15
+* Start bootloader from a read-only snapshot
+
+grub entries with GRUB_PARAM='ima_policy=tcb' and calling add_custom_grub_entries:
+* SLES 15
+* Advanced options for SLES 15
+* SLES 15 (ima_policy=tcb)
+* Advanced options for SLES 15 (ima_policy=tcb)
+* Start bootloader from a read-only snapshot
+
+And of course the new entries have ima_policy=tcb added to kernel parameters.
+
+=cut
+sub add_custom_grub_entries {
+    my $grub_param = get_var('GRUB_PARAM');
+    return unless defined($grub_param);
+
+    assert_script_run('cp /boot/grub2/grub.cfg grub.cfg.old');
+    upload_logs('grub.cfg.old', failok => 1);
+
+    assert_script_run('cp /etc/grub.d/40_custom 40_custom.tmp');
+    assert_script_run('cp /etc/grub.d/10_linux /etc/grub.d/11_linux_openqa');
+
+    my $cmd = "sed -i -e 's/\\(args=.\\)\\(\\\$4\\)/\\1$grub_param \\2/'";
+    $cmd .= " -e 's/\\(Advanced options for %s\\)/\\1 ($grub_param)/'";
+    $cmd .= " -e 's/\\(menuentry .\\\$(echo .\\\$title\\)/\\1 ($grub_param)/'";
+    $cmd .= " -e 's/\\(menuentry .\\\$(echo .\\\$os\\)/\\1 ($grub_param)/' /etc/grub.d/11_linux_openqa";
+    assert_script_run($cmd);
+    upload_logs('/etc/grub.d/11_linux_openqa', failok => 1);
+    assert_script_run('grub2-mkconfig -o /boot/grub2/grub.cfg');
+    upload_logs('/boot/grub2/grub.cfg', failok => 1);
+
+    my $distro = (is_sle() ? "SLES" : "openSUSE");
+    $distro .= " " if !is_pre_15();
+    $distro .= get_required_var('VERSION');
+    my $cnt_orig = script_output("grep -c 'menuentry .$distro' grub.cfg.old");
+    my $cnt_new  = script_output("grep -c 'menuentry .$distro' /boot/grub2/grub.cfg");
+    die("Unexpected number of grub entries: $cnt_new, expected: " . ($cnt_orig * 2)) if ($cnt_orig * 2 != $cnt_new);
+    $cnt_new = script_output("grep -c 'menuentry .$distro.*($grub_param)' /boot/grub2/grub.cfg");
+    die("Unexpected number of new grub entries: $cnt_new, expected: " . ($cnt_orig)) if ($cnt_orig != $cnt_new);
+    $cnt_new = script_output("grep -c 'linux.*/boot/vmlinu.* $grub_param ' /boot/grub2/grub.cfg");
+    die("Unexpected number of new grub entries with '$grub_param': $cnt_new, expected: " . ($cnt_orig)) if ($cnt_orig != $cnt_new);
+}
+
+=head2 boot_grub_item
+
+  boot_grub_item([ $menu1, [ $menu2 ] ]);
+
+Choose custom boot menu entry in grub.
+
+=cut
+sub boot_grub_item {
+    my ($self, $menu1, $menu2) = @_;
+    $menu1 = 1 unless defined($menu1);
+    $menu2 = 1 unless defined($menu2);
+    die((caller(0))[3] . " expects integer arguments ($menu1, $menu2") unless ($menu1 =~ /^\d+\z/ && $menu2 =~ /^\d+\z/);
+
+    assert_screen "grub2";
+
+    for my $i (1 .. $menu1) {
+        send_key "down";
+    }
+    sleep 1;
```

Comment:
Please avoid sleeps. Better wrap the send_key with `wait_screen_change` or just make it explicit with assert_screens which is often the better, more stable and even more time efficient option as the test can progress as soon as it has found the right screen. And multiple `sleep 1` still add up a lot of waiting time which can be saved by assert_screen :)

Reply:
I'll try `wait_screen_change`, but I want to avoid requesting users to change needles.